File: clients/capacitor/android/app/src/main/java/org/ubersdr/mobile/Http.java
package org.ubersdr.mobile;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.ConnectException;
import java.net.HttpURLConnection;
import java.net.SocketTimeoutException;
import java.net.URL;
import java.net.UnknownHostException;
import java.security.cert.CertPathValidatorException;
import java.security.cert.CertificateExpiredException;
import java.security.cert.X509Certificate;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLException;
import javax.net.ssl.SSLHandshakeException;
import javax.net.ssl.SSLPeerUnverifiedException;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

/**
 * One GET, returning JSON text — the whole HTTP surface this client needs
 * before a receiver is open.
 *
 * <p>This is the native half of clients/electron/discovery.js's {@code getJson},
 * and it exists for the same two reasons that one does not use Node's fetch: a
 * LAN receiver often has a self-signed certificate, which has to be refusable
 * by default and acceptable per receiver on an explicit click; and the caller is
 * not same-origin with any of these hosts, so from the WebView every one of
 * these would be a cross-origin request against a server whose CORS is an
 * operator setting that defaults off.
 *
 * <p>Failures are returned rather than thrown, classified into a code the
 * chooser can act on. The one that matters is {@code certError}: it is what
 * turns "this did not work" into an offer to trust the receiver anyway.
 */
final class Http {

    /** What came back, or why nothing did. */
    static final class Result {
        boolean ok;
        int status;
        String body;
        String code;
        String error;
        boolean certError;

        static Result success(int status, String body) {
            Result r = new Result();
            r.ok = true;
            r.status = status;
            r.body = body;
            return r;
        }

        static Result failure(String code, String error, boolean certError, int status) {
            Result r = new Result();
            r.ok = false;
            r.code = code;
            r.error = error;
            r.certError = certError;
            r.status = status;
            return r;
        }
    }

    // A response big enough to be a mistake. /api/instances with conditions is
    // a few hundred kilobytes; nothing this asks for is megabytes, and a socket
    // that keeps producing them should not be able to take the app down.
    private static final int MAX_BODY = 16 * 1024 * 1024;

    private Http() {}

    static Result getJson(String host, int port, boolean tls, boolean insecureTLS,
                          String path, int timeoutMs, String userAgent) {
        HttpURLConnection conn = null;
        try {
            // Bracketed for an IPv6 literal, which is otherwise unparseable as
            // an authority — the host:port colon and the address's own collide.
            String authority = host.indexOf(':') >= 0 ? "[" + host + "]" : host;
            URL url = new URL((tls ? "https://" : "http://") + authority + ":" + port + path);
            conn = (HttpURLConnection) url.openConnection();

            if (insecureTLS && conn instanceof HttpsURLConnection) {
                HttpsURLConnection https = (HttpsURLConnection) conn;
                https.setSSLSocketFactory(insecureFactory());
                // Scoped to this one connection: the receiver the operator
                // clicked "trust" for, and nothing else in the app.
                https.setHostnameVerifier((hostname, session) -> true);
            }

            conn.setRequestMethod("GET");
            conn.setRequestProperty("User-Agent", userAgent);
            conn.setRequestProperty("Accept", "application/json");
            conn.setConnectTimeout(timeoutMs);
            conn.setReadTimeout(timeoutMs);
            // Not followed, matching the desktop client, whose http.request does
            // not follow them either. It is not pedantry: a probe that followed
            // a redirect would report a row saying `tls: false, port: 80` for a
            // receiver actually served over https, and every address built from
            // that row afterwards — the saved entry, the window, the share link
            // — would be built from the address that did not answer.
            conn.setInstanceFollowRedirects(false);

            int status = conn.getResponseCode();
            if (status != 200) {
                drain(conn.getErrorStream());
                return Result.failure("HTTP_" + status, "HTTP " + status, false, status);
            }
            return Result.success(status, read(conn.getInputStream()));
        } catch (SocketTimeoutException e) {
            return Result.failure("ETIMEDOUT", "timeout", false, 0);
        } catch (UnknownHostException e) {
            return Result.failure("ENOTFOUND", "no such host", false, 0);
        } catch (ConnectException e) {
            return Result.failure("ECONNREFUSED", message(e, "connection refused"), false, 0);
        } catch (SSLPeerUnverifiedException e) {
            // The certificate is valid and is not this host's.
            return Result.failure("ERR_TLS_CERT_ALTNAME_INVALID", message(e, "certificate is for another host"), true, 0);
        } catch (SSLHandshakeException e) {
            return classifyHandshake(e);
        } catch (SSLException e) {
            return Result.failure("ERR_TLS_FAILED", message(e, "TLS failed"), true, 0);
        } catch (IOException e) {
            return Result.failure("EFAILED", message(e, "request failed"), false, 0);
        } catch (RuntimeException e) {
            // A malformed host from a directory row, most likely. Reported the
            // same way as anything else that did not answer.
            return Result.failure("EFAILED", message(e, "request failed"), false, 0);
        } finally {
            if (conn != null) conn.disconnect();
        }
    }

    /**
     * Which kind of certificate problem, in the vocabulary the desktop client
     * already uses (Node's OpenSSL codes), so the chooser shows the same words
     * on both. The distinction is cosmetic — every branch here is offered the
     * same "trust this receiver anyway" — but "certificate has expired" and
     * "self-signed certificate" are different things to be told.
     */
    private static Result classifyHandshake(SSLHandshakeException e) {
        for (Throwable t = e; t != null; t = t.getCause()) {
            if (t instanceof CertificateExpiredException) {
                return Result.failure("CERT_HAS_EXPIRED", "certificate has expired", true, 0);
            }
            if (t instanceof CertPathValidatorException) {
                return Result.failure("DEPTH_ZERO_SELF_SIGNED_CERT", "certificate is not trusted", true, 0);
            }
        }
        return Result.failure("UNABLE_TO_VERIFY_LEAF_SIGNATURE", message(e, "certificate could not be verified"), true, 0);
    }

    private static String message(Throwable e, String fallback) {
        String m = e.getMessage();
        return m == null || m.isEmpty() ? fallback : m;
    }

    private static String read(InputStream in) throws IOException {
        if (in == null) return "";
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] buf = new byte[16 * 1024];
        int n;
        int total = 0;
        while ((n = in.read(buf)) > 0) {
            total += n;
            if (total > MAX_BODY) throw new IOException("response too large");
            out.write(buf, 0, n);
        }
        in.close();
        return out.toString("UTF-8");
    }

    private static void drain(InputStream in) {
        if (in == null) return;
        try {
            byte[] buf = new byte[4096];
            while (in.read(buf) > 0) { /* to the end, so the socket can be reused */ }
        } catch (IOException ignored) {
            // Draining is an optimisation; the response is already decided.
        } finally {
            try { in.close(); } catch (IOException ignored) { }
        }
    }

    /**
     * A socket factory that accepts any certificate.
     *
     * <p>Built per call and applied to one connection, never installed as a
     * default: the operator asked to trust one receiver, and a factory left on
     * {@code HttpsURLConnection.setDefaultSSLSocketFactory} would quietly trust
     * everything the app talks to afterwards, including the directory.
     */
    private static javax.net.ssl.SSLSocketFactory insecureFactory() throws IOException {
        try {
            TrustManager[] trustAll = new TrustManager[]{
                new X509TrustManager() {
                    @Override public void checkClientTrusted(X509Certificate[] chain, String authType) { }
                    @Override public void checkServerTrusted(X509Certificate[] chain, String authType) { }
                    @Override public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
                }
            };
            SSLContext ctx = SSLContext.getInstance("TLS");
            ctx.init(null, trustAll, new java.security.SecureRandom());
            return ctx.getSocketFactory();
        } catch (java.security.GeneralSecurityException e) {
            throw new IOException("cannot build an insecure TLS context", e);
        }
    }
}

File: clients/capacitor/ios/UberSdrPlugin/sources/Http.swift
import Foundation

/// One GET, one JSON document, and the certificate-error classification the
/// chooser's "trust this receiver anyway" is built on.
///
/// The iOS half of `Http.java`, and deliberately the same shape: failures are
/// *returned* rather than thrown, classified into a code the chooser can act
/// on. The one that matters is `certError` — it is what turns "could not
/// connect" into an offer to trust a self-signed receiver, and a receiver on
/// somebody's home network very often has exactly that.
///
/// Why every call goes through here rather than `fetch` in the page, which is
/// the same reason as on Android and holds twice over on iOS: these are all
/// cross-origin from the app's own origin, against servers whose CORS defaults
/// off (`main.go`, `config.Server.EnableCORS`), so a chooser that fetched
/// `/api/description` itself would fail against most of the receivers it exists
/// to find. And a WKWebView cannot be told to accept a self-signed certificate
/// for one host, which is the whole point of `insecureTLS`.
enum Http {

    struct Result {
        var ok: Bool
        var status: Int = 0
        var body: String = ""
        var code: String = ""
        var error: String = ""
        var certError: Bool = false

        var payload: [String: Any] {
            // The keys discovery.js reads. Kept identical to the Android
            // plugin's so that src/discovery.js needs no idea which platform
            // answered it.
            [
                "ok": ok, "status": status, "body": body,
                "code": code, "error": error, "certError": certError
            ]
        }
    }

    /// A body larger than this is a receiver that is not a receiver. The same
    /// ceiling the Android side uses.
    private static let maxBody = 16 * 1024 * 1024

    static func getJson(host: String, port: Int, tls: Bool, insecureTLS: Bool,
                        path: String, timeoutMs: Int, userAgent: String,
                        completion: @escaping (Result) -> Void) {

        var components = URLComponents()
        components.scheme = tls ? "https" : "http"
        components.host = host
        // The default port is left out of the URL rather than written into it,
        // and this is not cosmetic. URLSession derives the Host header from the
        // URL, so an explicit :443 produces `Host: example.org:443` — which a
        // name-based virtual host does not match, and the directory answers 404
        // to. Every other client omits it (curl, and Java's HttpURLConnection,
        // which is why the Android half never met this), so a receiver that
        // works everywhere else would have appeared broken only here.
        let isDefaultPort = (tls && port == 443) || (!tls && port == 80)
        if !isDefaultPort { components.port = port }
        // `path` arrives complete with its query (`/api/instances?conditions=true`),
        // because that is how the shared discovery layer builds it. Assigning it
        // to `path` alone would percent-encode the '?', so it is parsed out.
        if let q = path.firstIndex(of: "?") {
            components.percentEncodedPath = String(path[path.startIndex..<q])
            components.percentEncodedQuery = String(path[path.index(after: q)...])
        } else {
            components.percentEncodedPath = path
        }

        guard let url = components.url else {
            completion(Result(ok: false, code: "EBADURL", error: "bad url"))
            return
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = Double(timeoutMs) / 1000.0
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        // A receiver that answers from a cache is a receiver whose free slots
        // and band conditions are wrong, which is most of what the chooser
        // draws.
        request.cachePolicy = .reloadIgnoringLocalCacheData

        // A session per call, because the delegate carries this call's answer to
        // the trust question and sessions outlive requests. Invalidated in the
        // completion so the delegate is released rather than retained forever.
        let delegate = TrustDelegate(insecureTLS: insecureTLS)
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = Double(timeoutMs) / 1000.0
        config.httpCookieStorage = nil
        let session = URLSession(configuration: config, delegate: delegate, delegateQueue: nil)

        let task = session.dataTask(with: request) { data, response, error in
            defer { session.finishTasksAndInvalidate() }

            if let error = error as NSError? {
                completion(classify(error, sawCertFailure: delegate.sawCertFailure))
                return
            }
            guard let http = response as? HTTPURLResponse else {
                completion(Result(ok: false, code: "EFAILED", error: "no response"))
                return
            }
            guard http.statusCode == 200 else {
                completion(Result(ok: false, status: http.statusCode,
                                  code: "HTTP_\(http.statusCode)",
                                  error: "HTTP \(http.statusCode)"))
                return
            }
            let bytes = data ?? Data()
            guard bytes.count <= maxBody else {
                completion(Result(ok: false, code: "ETOOBIG", error: "response too large"))
                return
            }
            completion(Result(ok: true, status: 200,
                              body: String(decoding: bytes, as: UTF8.self)))
        }
        task.resume()
    }

    /// URLError codes into the vocabulary the chooser already speaks.
    ///
    /// The names are Node's OpenSSL/libuv ones, because the desktop client
    /// produces those and the chooser — which is the same page in both clients —
    /// shows them. A code invented here would be a string no other client can
    /// produce and no message in the chooser knows about.
    private static func classify(_ error: NSError, sawCertFailure: Bool) -> Result {
        let certCodes: Set<Int> = [
            NSURLErrorServerCertificateUntrusted,
            NSURLErrorServerCertificateHasBadDate,
            NSURLErrorServerCertificateHasUnknownRoot,
            NSURLErrorServerCertificateNotYetValid,
            NSURLErrorSecureConnectionFailed,
            NSURLErrorClientCertificateRejected
        ]

        if sawCertFailure || certCodes.contains(error.code) {
            let code: String
            switch error.code {
            case NSURLErrorServerCertificateHasUnknownRoot,
                 NSURLErrorServerCertificateUntrusted:
                code = "DEPTH_ZERO_SELF_SIGNED_CERT"
            case NSURLErrorServerCertificateHasBadDate,
                 NSURLErrorServerCertificateNotYetValid:
                code = "CERT_HAS_EXPIRED"
            default:
                code = "ERR_TLS_CERT_ALTNAME_INVALID"
            }
            return Result(ok: false, code: code,
                          error: error.localizedDescription, certError: true)
        }

        let code: String
        switch error.code {
        case NSURLErrorTimedOut:              code = "ETIMEDOUT"
        case NSURLErrorCannotFindHost:        code = "ENOTFOUND"
        case NSURLErrorCannotConnectToHost:   code = "ECONNREFUSED"
        case NSURLErrorNetworkConnectionLost: code = "ECONNRESET"
        case NSURLErrorNotConnectedToInternet: code = "ENETUNREACH"
        default:                              code = "EFAILED"
        }
        return Result(ok: false, code: code, error: error.localizedDescription)
    }

    /// Per-receiver certificate trust, and nothing wider.
    ///
    /// `insecureTLS` is a decision the operator made about one receiver in the
    /// chooser, so it is scoped to one session serving one request. The tempting
    /// alternatives — a global ATS exception for TLS validation, or a shared
    /// session with this delegate — would silently extend that decision to every
    /// receiver, which is exactly what the per-entry flag exists to prevent.
    private final class TrustDelegate: NSObject, URLSessionDelegate {
        let insecureTLS: Bool
        /// Set when a challenge was refused, so `classify` can report a
        /// certificate failure even where URLSession reports only the generic
        /// "secure connection failed" that follows it.
        private(set) var sawCertFailure = false

        init(insecureTLS: Bool) {
            self.insecureTLS = insecureTLS
        }

        func urlSession(_ session: URLSession,
                        didReceive challenge: URLAuthenticationChallenge,
                        completionHandler: @escaping (URLSession.AuthChallengeDisposition,
                                                      URLCredential?) -> Void) {
            guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
                  let trust = challenge.protectionSpace.serverTrust else {
                completionHandler(.performDefaultHandling, nil)
                return
            }
            if insecureTLS {
                completionHandler(.useCredential, URLCredential(trust: trust))
                return
            }
            // Not "reject": let the system evaluate it, and only note that the
            // failure was a trust one so the chooser can offer the override.
            var err: CFError?
            if SecTrustEvaluateWithError(trust, &err) {
                completionHandler(.useCredential, URLCredential(trust: trust))
            } else {
                sawCertFailure = true
                completionHandler(.cancelAuthenticationChallenge, nil)
            }
        }
    }
}
